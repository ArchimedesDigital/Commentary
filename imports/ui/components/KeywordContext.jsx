import { Meteor } from 'meteor/meteor';
import React from 'react';
import RaisedButton from 'material-ui/RaisedButton';
import { queryCommentWithKeyword, makeKeywordContextQueryFromComment } from '../../api/api_utils.js';

export default KeywordContext = React.createClass({

	mixins: [ReactMeteorData],

	propTypes: {
		slug: React.PropTypes.string.isRequired,
		maxLines: React.PropTypes.number.isRequired,
	},

	getInitialState() {
		return {
			selectedLemmaEdition: {
				lines: [],
			},
		}
	},

	getMeteorData() {
		let lemmaText = [];

		const commentsSub = Meteor.subscribe('comments.keyword_context', this.props.slug);

		if (commentsSub.ready()) {
			const commentCursor = queryCommentWithKeyword(this.props.slug);
			if (commentCursor.count() > 0) {
				const comment = commentCursor.fetch()[0];

				const textNodesQuery = makeKeywordContextQueryFromComment(comment, this.props.maxLines);
				const textNodesSub = Meteor.subscribe('textnodes.keyword_context', textNodesQuery);

				if (textNodesSub.ready()) {
					const textNodesCursor = TextNodes.find(textNodesQuery);
					lemmaText = this.textFromTextNodesGroupedByEdition(textNodesCursor);
				}
			}
		}

		return {
			lemmaText,
		};
	},

	textFromTextNodesGroupedByEdition(nodesCursor) {
		const editions = [];
		nodesCursor.forEach(node => {
			node.text.forEach(text => {
				let myEdition = editions.find((e) => {
					return text.edition.slug === e.slug;
				});

				if (!myEdition) {
					myEdition = {
						title: text.edition.title,
						slug: text.edition.slug,
						lines: [],
					}
					editions.push(myEdition);
				}

				myEdition.lines.push({
					html: text.html,
					n: text.n,
				});
			});
		});

		return editions;
	},

	toggleEdition(editionSlug){
		if(this.state.selectedLemmaEdition.slug !== editionSlug){
			var newSelectedEdition = {};
			this.data.lemmaText.forEach(function(edition){
				if(edition.slug === editionSlug){
					newSelectedEdition = edition;
				}
			});

			this.setState({
				selectedLemmaEdition: newSelectedEdition
			});
		}
	},

	componentDidUpdate(){
		if (this.data.lemmaText.length > 0 && this.state.selectedLemmaEdition.lines.length === 0) {
			this.setState({
				selectedLemmaEdition: this.data.lemmaText[0]
			});
		}
	},

	render() {
		return (<article className="comment lemma-comment paper-shadow">
			{
				this.state.selectedLemmaEdition.lines.map((line, i) => {
					return <p
						key={i}
						className="lemma-text"
						dangerouslySetInnerHTML={{__html: line.html}}
					/>;
				})
			}
			<div className="edition-tabs tabs">
				{
					this.data.lemmaText.map((lemmaTextEdition, i) => {
						let lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

						return <RaisedButton
							key={i}
							label={lemmaEditionTitle}
							data-edition={lemmaTextEdition.title}
							className={this.state.selectedLemmaEdition.slug === lemmaTextEdition.slug ? "edition-tab tab selected-edition-tab" : "edition-tab tab"}
							onClick={this.toggleEdition.bind(null, lemmaTextEdition.slug)}
						>
						</RaisedButton>
					})
				}
			</div>
		</article>);
	},

});
