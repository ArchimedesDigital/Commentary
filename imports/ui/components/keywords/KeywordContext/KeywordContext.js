import React from 'react';
import { Meteor } from 'meteor/meteor';
import { Session } from 'meteor/session';
import { createContainer } from 'meteor/react-meteor-data';

import getMuiTheme from 'material-ui/styles/getMuiTheme';
import RaisedButton from 'material-ui/RaisedButton';
import FontIcon from 'material-ui/FontIcon';

// lib
import muiTheme from '/imports/lib/muiTheme';
import { queryCommentWithKeywordId, makeKeywordContextQueryFromComment } from '/imports/lib/utils';


const KeywordContext = React.createClass({

	propTypes: {
		keyword: React.PropTypes.object,
		lemmaText: React.PropTypes.array,
		context: React.PropTypes.object,
	},

	childContextTypes: {
		muiTheme: React.PropTypes.object.isRequired,
	},

	getInitialState() {
		return {
			selectedLemma: 0,
		};
	},

	getChildContext() {
		return { muiTheme: getMuiTheme(muiTheme) };
	},

	textFromTextNodesGroupedByEdition(nodesCursor) {
		const editions = [];
		nodesCursor.forEach(node => {
			node.text.forEach(text => {
				let myEdition = editions.find((e) => text.edition.slug === e.slug);

				if (!myEdition) {
					myEdition = {
						title: text.edition.title,
						slug: text.edition.slug,
						lines: [],
					};
					editions.push(myEdition);
				}

				myEdition.lines.push({
					html: text.html,
					n: text.n,
				});
			});
		});

		// sort lines for each edition by line number
		for (let i = 0; i < editions.length; ++i) {
			editions[i].lines.sort((a, b) => {
				if (a.n < b.n) {
					return -1;
				} else if (b.n < a.n) {
					return 1;
				} else {
					return 0;
				}
			});
		}

		return editions;
	},

	toggleEdition(newSelectedLemma) {
		if (this.state.selectedLemma !== newSelectedLemma && newSelectedLemma < this.props.lemmaText.length) {
			this.setState({
				selectedLemma: newSelectedLemma,
			});
		}
	},

	render() {
		const { keyword, context, lemmaText } = this.props;

		if (!lemmaText || !lemmaText.length) {
			return null;
		}

		return (
			<article className="comment lemma-comment paper-shadow keyword-context">
				<div>
					<span className="lemma-comment-ref-header">
						{keyword.work.title} {keyword.subwork.n}.{keyword.lineFrom}{(keyword.lineTo && keyword.lineFrom !== keyword.lineTo) ? `-${keyword.lineTo}` : ''}
					</span>
					{lemmaText[this.state.selectedLemma].lines.map((line, i) =>
						<p
							key={i}
							className="lemma-text"
							dangerouslySetInnerHTML={{ __html: line.html }}
						/>
					)}
				</div>
				<div className="edition-tabs tabs">
					{
						lemmaText.map((lemmaTextEdition, i) => {
							const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

							return (
								<RaisedButton
									key={i}
									label={lemmaEditionTitle}
									data-edition={lemmaTextEdition.title}
									className={this.state.selectedLemma === i ? 'edition-tab tab selected-edition-tab' : 'edition-tab tab'}
									onClick={this.toggleEdition.bind(null, i)}
								/>
							);
						})
					}
				</div>
				<div className="context-tabs tabs">
					<RaisedButton
						className="context-tab tab"
						href={`/commentary/?works=${context.work}&subworks=${context.subwork
						}&lineFrom=${context.lineFrom}&lineTo=${context.lineTo}`}
						label="Context"
						labelPosition="before"
						icon={<FontIcon className="mdi mdi-chevron-right" />}
					/>
				</div>
			</article>
		);
	},

});

const KeywordContextContainer = createContainer(({ keyword, maxLines }) => {
	let lemmaText = [];
	const context = {};

	if (!keyword) {
		return {
			lemmaText,
			context,
		};
	}

	if (keyword.work && keyword.subwork && keyword.lineFrom) {
		const textNodesQuery = {
			'work.slug': keyword.work.slug,
			'subwork.n': keyword.subwork.n,
			'text.n': {
				$gte: keyword.lineFrom,
				$lte: keyword.lineFrom,
			},
		};
		if (keyword.lineTo) {
			textNodesQuery['text.n'].$lte = keyword.lineTo;
		}
		const textNodesSub = Meteor.subscribe('textnodes.keyword_context', textNodesQuery);

		context.work = textNodesQuery['work.slug'];
		context.subwork = textNodesQuery['subwork.n'];
		context.lineFrom = textNodesQuery['text.n'].$gte;
		context.lineTo = textNodesQuery['text.n'].$lte;

		if (textNodesSub.ready()) {
			const textNodesCursor = TextNodes.find(textNodesQuery);
			lemmaText = this.textFromTextNodesGroupedByEdition(textNodesCursor);
		}

	} else {
		const commentsSub = Meteor.subscribe('comments.keyword_context', keyword._id, Session.get('tenantId'));

		if (commentsSub.ready()) {
			const commentCursor = queryCommentWithKeywordId(keyword._id);

			if (commentCursor.count() > 0) {
				const comment = commentCursor.fetch()[0];
				const textNodesQuery = makeKeywordContextQueryFromComment(comment, maxLines);
				const textNodesSub = Meteor.subscribe('textnodes.keyword_context', textNodesQuery);

				context.work = textNodesQuery['work.slug'];
				context.subwork = textNodesQuery['subwork.n'];
				context.lineFrom = textNodesQuery['text.n'].$gte;
				context.lineTo = textNodesQuery['text.n'].$lte;

				if (textNodesSub.ready()) {
					const textNodesCursor = TextNodes.find(textNodesQuery);
					lemmaText = this.textFromTextNodesGroupedByEdition(textNodesCursor);
				}
			}
		}
	}

	return {
		lemmaText,
		context,
	};
}, KeywordContext);


export default KeywordContextContainer;
