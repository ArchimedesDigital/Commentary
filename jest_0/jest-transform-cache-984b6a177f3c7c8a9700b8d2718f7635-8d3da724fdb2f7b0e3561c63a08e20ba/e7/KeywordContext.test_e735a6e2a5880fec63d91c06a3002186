/* istanbul ignore next */'use strict';var /* istanbul ignore next */_react = require('react'); /* istanbul ignore next */var _react2 = _interopRequireDefault(_react);
var /* istanbul ignore next */_reactTestRenderer = require('react-test-renderer'); /* istanbul ignore next */var _reactTestRenderer2 = _interopRequireDefault(_reactTestRenderer);


var /* istanbul ignore next */_KeywordContext = require('./KeywordContext'); /* istanbul ignore next */var _KeywordContext2 = _interopRequireDefault(_KeywordContext);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

describe('KeywordContext', function () {
	it('renders correctly', function () {

		var tree = /* istanbul ignore next */_reactTestRenderer2.default.
		create( /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_KeywordContext2.default, /* istanbul ignore next */null)).
		toJSON();
		expect(tree).toMatchSnapshot();
	});
}); // component: