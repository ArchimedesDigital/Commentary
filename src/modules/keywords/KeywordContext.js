import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { compose } from 'react-apollo';
import RaisedButton from 'material-ui/RaisedButton';
import FontIcon from 'material-ui/FontIcon';

// lib
import Utils, { makeKeywordContextQueryFromComment } from '../../lib/utils';

// graphql
import { editionsQuery } from '../../graphql/methods/editions';
import { textNodesQuery } from '../../graphql/methods/textNodes';
import { commentsQuery } from '../../graphql/methods/comments';

class KeywordContext extends Component {
	constructor(props) {
		super(props);
		this.state = {
			selectedLemma: 0,
		};
		this.toggleEdition = this.toggleEdition.bind(this);
	}
	componentWillReceiveProps(props) {
		if (props.commentsQuery.loading ||
			props.editionsQuery.loading ||
			props.textNodesQuery.loading) {
			return;
		}

		const { keyword, maxLines } = props;
		const tenantId = sessionStorage.getItem('tenantId');

		let lemmaText = [];
		const context = {};

		if (!keyword) {
			return {
				lemmaText,
				context,
			};
		}

		if (keyword.work && keyword.subwork && keyword.lineFrom) {


			context.work = keyword.work.slug;
			context.subwork = keyword.subwork.n;
			context.lineFrom = keyword.lineFrom;
			context.lineTo = keyword.lineTo ? keyword.lineTo : keyword.lineFrom;

			if (!props.textNodesQuery.loading) {
				const textNodesCursor = props.textNodesQuery.textNodes;
				lemmaText = Utils.textFromTextNodesGroupedByEdition(textNodesCursor, props.editionsQuery.editions);
			}
			props.textNodesQuery.refetch({
				lineFrom: context.lineFrom,
				lineTo: context.lineTo,
				subworkN: context.subwork,
				workSlug: context.work
			});

		} else {
			if (tenantId) {
				const query = {};
				query['keyword._id'] = keyword._id;
				props.commentsQuery.refetch({
					queryParam: JSON.stringify(query),
					limit: 1
				});
			}
			
			if (!props.commentsQuery.loading) {

				if (props.commentsQuery.comments.length > 0) {
					const comment = props.commentsQuery.comments[0];
					const query = makeKeywordContextQueryFromComment(comment, maxLines);
					props.textNodesQuery.refetch(query);
					context.work = query.workSlug;
					context.subwork = query.subworkN;
					context.lineFrom = query.lineFrom;
					context.lineTo = query.lineTo;

					const textNodesCursor = textNodesQuery.loading ? [] : textNodesQuery.textNodes;
					lemmaText = Utils.textFromTextNodesGroupedByEdition(textNodesCursor, props.editionsQuery.editions);
				}
			}
		}
		this.setState({
			lemmaText: lemmaText,
			context: context,
		});
	}
	toggleEdition(newSelectedLemma) {
		if (this.state.selectedLemma !== newSelectedLemma && newSelectedLemma < this.state.lemmaText.length) {
			this.setState({
				selectedLemma: newSelectedLemma,
			});
		}
	}

	render() {
		const { keyword} = this.props;
		const { context, lemmaText } = this.state;

		if (!lemmaText || !lemmaText.length) {
			return null;
		}

		return (
			<article className="comment lemma-comment paper-shadow keyword-context">
				<div>
					<span className="lemma-comment-ref-header">
						{keyword.work.title} {keyword.subwork.n}.{keyword.lineFrom}{(keyword.lineTo && keyword.lineFrom !== keyword.lineTo) ? `-${keyword.lineTo}` : ''}
					</span>
					{lemmaText[this.state.selectedLemma].lines.map((line, i) =>
						(<p
							key={i}
							className="lemma-text"
							dangerouslySetInnerHTML={{ __html: line.html }}
						/>)
					)}
				</div>
				<div className="edition-tabs tabs">
					{
						lemmaText.map((lemmaTextEdition, i) => {
							const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

							return (
								<RaisedButton
									key={i}
									label={lemmaEditionTitle}
									data-edition={lemmaTextEdition.title}
									className={this.state.selectedLemma === i ? 'edition-tab tab selected-edition-tab' : 'edition-tab tab'}
									onClick={this.toggleEdition.bind(null, i)}
								/>
							);
						})
					}
				</div>
				<div className="context-tabs tabs">
					<RaisedButton
						className="context-tab tab"
						href={`/commentary/?works=${context.work}&subworks=${context.subwork
						}&lineFrom=${context.lineFrom}&lineTo=${context.lineTo}`}
						label="Context"
						labelPosition="before"
						icon={<FontIcon className="mdi mdi-chevron-right" />}
					/>
				</div>
			</article>
		);
	}

}
KeywordContext.propTypes = {
	keyword: PropTypes.object,
	editionsQuery: PropTypes.object,
	commentsQuery: PropTypes.object,
	textNodesQuery: PropTypes.object,
	maxLines: PropTypes.number
};

export default compose(
	editionsQuery,
	commentsQuery,
	textNodesQuery
)(KeywordContext);
