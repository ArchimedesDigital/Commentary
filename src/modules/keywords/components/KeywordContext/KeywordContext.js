import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { compose } from 'react-apollo';
import RaisedButton from 'material-ui/RaisedButton';
import FontIcon from 'material-ui/FontIcon';

// lib
import Utils, { makeKeywordContextQueryFromComment } from '../../../../lib/utils';

// graphql
import commentsQuery from '../../../comments/graphql/queries/comments';
import { editionsQuery } from '../../../../graphql/methods/editions';
import { textNodesQuery } from '../../../../graphql/methods/textNodes';

class KeywordContext extends Component {
	constructor(props) {
		super(props);
		this.state = {
			selectedLemma: 0,
		};
		this.toggleEdition = this.toggleEdition.bind(this);
	}
	componentWillReceiveProps(props) {
		if (props.commentsQuery.loading ||
			props.editionsQuery.loading ||
			props.textNodesQuery.loading) {
			return;
		}

		const { keyword, maxLines } = props;
		const tenantId = sessionStorage.getItem('tenantId');

		let lemmaText = [];
		const context = {};

		if (!keyword) {
			return {
				lemmaText,
				context,
			};
		}

		if (keyword.work && keyword.subwork && keyword.lineFrom
			&& !props.textNodesQuery.variables.workUrn && !props.textNodesQuery.variables.work) {


			context.work = keyword.work.slug;
			context.subwork = keyword.subwork.n;
			context.lineFrom = keyword.lineFrom;
			context.lineTo = keyword.lineTo ? keyword.lineTo : keyword.lineFrom;
			if (!props.textNodesQuery.loading) {
				const textNodesCursor = props.textNodesQuery.collections[0].textGroups[0].works;
				lemmaText = Utils.textFromTextNodesGroupedByEdition(textNodesCursor, props.editionsQuery.collections[0].textGroups[0].works);
			}
			const properties = Utils.getUrnTextNodesProperties(Utils.createLemmaCitation(
				context.work, context.lineFrom, context.lineTo
			));
			props.textNodesQuery.refetch(properties);
			return;

		} else {
			if (tenantId) {
				const query = {};
				query['keyword._id'] = keyword._id;
				props.commentsQuery.refetch({
					queryParam: JSON.stringify(query),
					limit: 1
				});
			}

			if (props.commentsQuery.comments.length > 0) {
				const comment = props.commentsQuery.comments[0];
				const query = makeKeywordContextQueryFromComment(comment, maxLines); //TODO
				context.work = query.workSlug;
				context.subwork = query.subworkN;
				context.lineFrom = query.lineFrom;
				context.lineTo = query.lineTo;
			}
			const textNodesCursor = props.textNodesQuery.collections[0].textGroups[0].works;

			lemmaText = Utils.textFromTextNodesGroupedByEdition(textNodesCursor, props.editionsQuery.collections[0].textGroups[0].works);
		}
		this.setState({
			lemmaText: lemmaText,
			context: context,
		});
	}
	toggleEdition(newSelectedLemma) {
		if (this.state.selectedLemma !== newSelectedLemma && newSelectedLemma < this.state.lemmaText.length) {
			this.setState({
				selectedLemma: newSelectedLemma,
			});
		}
	}

	render() {
		const { keyword} = this.props;
		const { context, lemmaText } = this.state;

		if (!lemmaText || !lemmaText.length) {
			return null;
		}

		return (
			<article className="comment lemma-comment paper-shadow keyword-context">
				<div>
					<span className="lemma-comment-ref-header">
						{keyword.work.title} {keyword.subwork.n}.{keyword.lineFrom}{(keyword.lineTo && keyword.lineFrom !== keyword.lineTo) ? `-${keyword.lineTo}` : ''}
					</span>
					{lemmaText[this.state.selectedLemma].lines.map((line, i) =>
						(<p
							key={i}
							className="lemma-text"
							dangerouslySetInnerHTML={{ __html: line.html }}
						/>)
					)}
				</div>
				<div className="version-tabs tabs">
					{
						lemmaText.map((lemmaTextEdition, i) => {
							const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

							return (
								<RaisedButton
									key={i}
									label={lemmaEditionTitle}
									data-edition={lemmaTextEdition.title}
									className={this.state.selectedLemma === i ? 'version-tab tab selected-version-tab' : 'version-tab tab'}
									onClick={this.toggleEdition.bind(null, i)}
								/>
							);
						})
					}
				</div>
				<div className="context-tabs tabs">
					<RaisedButton
						className="context-tab tab"
						href={`/commentary/?works=${context.work}&subworks=${context.subwork
						}&lineFrom=${context.lineFrom}&lineTo=${context.lineTo}`}
						label="Context"
						labelPosition="before"
						icon={<FontIcon className="mdi mdi-chevron-right" />}
					/>
				</div>
			</article>
		);
	}

}
KeywordContext.propTypes = {
	keyword: PropTypes.object,
	editionsQuery: PropTypes.object,
	commentsQuery: PropTypes.object,
	textNodesQuery: PropTypes.object,
	maxLines: PropTypes.number
};

export default compose(
	editionsQuery,
	commentsQuery,
	textNodesQuery
)(KeywordContext);
