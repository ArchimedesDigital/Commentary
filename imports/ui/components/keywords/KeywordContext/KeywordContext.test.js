import React from 'react';
import renderer from 'react-test-renderer';

// component:
import KeywordContext from './KeywordContext';

describe('KeywordContext', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(<KeywordContext />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
