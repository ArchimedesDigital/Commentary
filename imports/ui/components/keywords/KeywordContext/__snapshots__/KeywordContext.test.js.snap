// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeywordContext renders correctly 1`] = `null`;
